File: Array/linchuanxin/Algorithm/day65_0804_leetcode405_数字转换为十六进制数/day65_leetcode405.c
//
// Created by LCX on 2020/8/6.
//
#include <stdio.h>
#include <string.h>
#include <malloc.h>

char * toHex(int num){
    char *s = "0123456789abcdef";
    char *index = (char*)malloc(strlen(s));
    unsigned int num1 = num;
    int top = -1;
    if(num1 == 0)
        return "0";
    while (num1 != 0)
    {
        int tmp = num1 & 15;
        index[++top] = s[tmp];
        num1 = num1 >> 4;

    }
    char *str = (char*)malloc(top+1);
    for (int i = 0; i <= top+1; ++i) {
        //printf("%c",index[top]);
        //str = str + index[top];
        str[i] = index[top];
        top--;
    }
    return str;
}


//  8    1 0  0  0
//       1 1  1  1
//       1 0  0  0
int main()
{
    //toHex(26);
    printf("to hex: %s",toHex(8));
    return 0;
}

File: Array/linchuanxin/Algorithm/day65_0804_leetcode405_数字转换为十六进制数/day65_leetcode405.java
package day65_0804_leetcode405;



public class day65_leetcode405 {

    public static String toHex(int num) {
        StringBuilder res = new StringBuilder();
        char[] arr = "0123456789abcdef".toCharArray();
        if(num == 0)
            return "0";
        while (num != 0){
            int tmp = num & 15;
            res.append(arr[tmp]);
            num = num >>> 4;
        }
        return res.reverse().toString();
    }

    public static void main(String[] args) {
        System.out.println(day65_leetcode405.toHex(26));
    }


}
